s3-encryption/client.md

```diff
@@ -0,0 +1,142 @@
+## Overview
+
+This document describes the top-level public S3 Encryption Client (S3EC).
+The S3EC provides client-side encryption for Amazon S3.
+
+## AWS SDK Compatibility
+
+The S3EC MUST adhere to the same interface for API operations as the conventional AWS SDK S3 client.
+In other words, the SDK's conventional S3 client is able to be substituted for the S3EC.
+The S3EC SHOULD support invoking operations unrelated to client-side encryption e.g. CopyObject as the conventional AWS SDK S3 client would.
+The S3EC MUST provide a different set of configuration options than the conventional S3 client.
+
+## Configuration
+
+### Cryptographic Materials
+
+The S3EC MUST accept either one CMM or one Keyring instance upon initialization.
+If both a CMM and a Keyring are provided, the S3EC MUST throw an exception.
+When a Keyring is provided, the S3EC MUST create an instance of the DefaultCMM using the provided Keyring.
+
+The S3EC MAY accept key material directly.
+When only key material is provided, a Keyring corresponding to the type of key material is created by default.
+This behavior is discouraged, as it requires all Keyring configuration options to be supported by client initialization.
+This leads to customer confusion when a Keyring is provided and a Keyring option is set on the client, and thus not applied.
+It is considered deprecated, meaning it will be removed in the next major version (v4).
+
+### Options
+
+#### Enable Legacy Wrapping Algorithms
+
+The S3EC MUST support the option to enable or disable legacy wrapping algorithms.
+The option to enable legacy wrapping algorithms MUST be set to false by default.
+When enabled, the S3EC MUST be able to decrypt objects encrypted with all supported wrapping algorithms (both legacy and fully supported).
+When disabled, the S3EC MUST NOT decrypt objects encrypted using legacy wrapping algorithms; it MUST throw an exception when attempting to decrypt an object encrypted with a legacy wrapping algorithm.
+
+#### Enable Legacy Unauthenticated Modes
+
+The S3EC MUST support the option to enable or disable legacy unauthenticated modes (content encryption algorithms).
+The option to enable legacy unauthenticated modes MUST be set to false by default.
+When enabled, the S3EC MUST be able to decrypt objects encrypted with all content encryption algorithms (both legacy and fully supported).
+When disabled, the S3EC MUST NOT decrypt objects encrypted using legacy content encryption algorithms; it MUST throw an exception when attempting to decrypt an object encrypted with a legacy content encryption algorithm.
+
+#### Enable Delayed Authentication
+
+The S3EC MUST support the option to enable or disable Delayed Authentication mode.
+Delayed Authentication mode MUST be set to false by default.
+When enabled, the S3EC MAY release plaintext from a stream which has not been authenticated.
+When disabled the S3EC MUST NOT release plaintext from a stream which has not been authenticated.
+
+#### Set Buffer Size
```

Comment:
what should happen if set buffer size is enabled but delayed auth mode is enabled? There is no behavior definition for this configuration? error? happy?

Reply:
good shout, details added